### ISSUE TITLE
DC_DOUBLECHECK: double-checked locking no longer detected when the re-read of the field is routed through an identity call

### ISSUE THREAD

**Issue by MarkLee131** [spotbugs/spotbugs#4212]
**Describe the bug**

`DC_DOUBLECHECK` reports double-checked locking on a non-volatile field. The re-read of the field inside the synchronized block, `z = x`, can be routed through `Collections.singletonList(x).get(0)`, and the detector stops reporting, although the double-checked locking idiom and its unsafe publication are completely unchanged.

`IS2_INCONSISTENT_SYNC` is reported on both files and serves as an internal control: the file is still being analyzed, and the detector that does not depend on the rewritten expression is unaffected.

**To Reproduce**

1. Save the two files below.
2. Compile each and run SpotBugs on the resulting classes:

```
javac -g -d classes/ DclSeed.java DclVariant.java
spotbugs -textui -effort:max -low -xml:withMessages -output out.xml classes/

# XML rather than the default text output: -textui prints SHORT bug
# codes (e.g. "EI", "DC"), while out.xml carries the published
# pattern ids this report quotes. Both classes go in one run; SpotBugs
# analyzes bytecode per class, so there is no cross-file interference.
```

3. Compare the reported bug patterns.

`DclSeed.java`:

```java
public class DclSeed {
    Object x;

    void m() {
        Object z = x;
        if (z == null) {
            synchronized (this) {
                z = x;                                    // DC_DOUBLECHECK reported
                if (z == null) {
                    x = z = new Object();
                }
            }
        }
    }
}
```

`DclVariant.java`:

```java
public class DclVariant {
    Object x;

    void m() {
        Object z = x;
        if (z == null) {
            synchronized (this) {
                z = java.util.Collections.singletonList(x).get(0);   // same reference; not reported
                if (z == null) {
                    x = z = new Object();
                }
            }
        }
    }
}
```

**Expected behavior**

`DC_DOUBLECHECK` on `DclVariant.m()` as well, since the locking structure and the field's non-volatility are identical.

The two programs are behaviourally identical, so the same bug pattern should be reported on both.

**Environment**

- Java Compiler: Eclipse Temurin
- Java version: JDK 25 (25.0.3+9)
- SpotBugs Version: 4.10.3
- How do you use SpotBugs: CLI (`spotbugs -textui`)

**Comment by adityaanikam** [comment]
Traced the exact reason `DclVariant` slips through. `FindDoubleCheck.java`'s stage 2 only reports the bug if the null check lands within `countSinceGetReference < 5` opcodes of the reload inside the synchronized block. `z = x` stays well under that. `z = Collections.singletonList(x).get(0)` adds enough extra bytecode, the static call, the interface call, the extra load for the comparison, to push past the window, so stage 2 never fires and the detector falls back to stage 0. The two reproducers you gave make the mechanism completely legible. Happy to widen that window or track the reload with real dataflow instead of counting opcodes, want a specific direction before I start, or should I just pick one?

**Comment by MarkLee131** [comment]
Thanks, that matches my read.

I would skip widening the window. Any threshold can be pushed past by a longer wrapper, and the window never checks that the tested value comes from the reload. Strict dataflow has the opposite issue. Value tracking cannot see that `get(0)` returns `x`, so an aliasing rule would still miss this variant.

A structural rule avoids both. It requires a `GETFIELD` of the field inside the synchronized region, plus a null guarded `PUTFIELD` of the same field there, with no distance limit. The `PUTFIELD` is the real sink, so the origin of the tested value stops mattering.
